### ISSUE TITLE
RESTEasy Reactive: form input uses different kinds of unserialisers

### ISSUE THREAD

**Issue by FroMage** [quarkusio/quarkus#22210]
### Describe the bug

If I have `@RestForm` parameters in my endpoint, they use `ParamConverter`, but if I do like https://quarkus.io/guides/resteasy-reactive#handling-multipart-form-data tells me and move them to a wrapper class, they use `MessageBodyReader`.

I'm pretty sure that this is due to multipart, because bean classes should not change how stuff are deserialised.

Is this by design or by accident? If it's by accident, perhaps we need to fix it. If by design, we should probably document it because it's not obvious at all, and also it requires me to write silly endpoints like these:

```java
@Path("albums")
public class Albums extends MyController {
	
    public static class MyForm {
        @RestForm @Length(max = Util.VARCHAR_SIZE) String title;
        @RestForm String photoUrl;
        @RestForm boolean removePhoto;
        @RestForm File photo;
    }

    @POST
    @Consumes(MediaType.MULTIPART_FORM_DATA)
    @Authenticated
	public void edit(@RestPath long id,
	        MyForm myForm,
                // notice how this one won't fit in MyForm because I need a custom ParamConverter
                @RestForm Date releaseDate){
	}
}
```

**Comment by geoand** [comment]
> I'm pretty sure that this is due to multipart

Yes, this is correct. We treat each part of the multipart request as a body

**Comment by FroMage** [comment]
Well, this is very confusing from the PoV of users because a switch from `urlencoded` to `multipart` caused my `@FormParam` to stop being read, due to my having to create a `MessageBodyReader<Date>`.

Why is there this difference? Can we have parts which are entities? Normally entities (bodies) don't have a `@FormParam` annotation. The spec does say that anything annotated with `@FormParam` should use `ParamConverter`.

**Comment by geoand** [comment]
The way I see it, a form param of a urlencoded body is like a query param or a path param and for multipart, each part is like a message body.

**Comment by FroMage** [comment]
Well. This is sorta true. But it's also confusing, Perhaps we can mitigate this by having multipart fallback to `ParamConverter` if it can't find a good `MessageBodyReader`?

**Comment by geoand** [comment]
Yeah, that sounds reasonable

**Comment by jnbdz** [comment]
Checked this on current main. The difference described here is gone: a `@RestForm` parameter and the same field moved into a form class both go through the `ParamConverter` now, for `multipart/form-data` as well as `application/x-www-form-urlencoded`. That came with the multipart rework in [03833f802b4](https://github.com/quarkusio/quarkus/commit/03833f802b4e2949e010d3e3850284aa04812b2c) (#22205), which uses a converter unless a part media type is declared, so the `Date` example from the description works as a parameter and as a form class field.

What is left is the mitigation suggested above, the fallback when a reader is missing. With `@PartType` set to a media type that no `MessageBodyReader` handles, the request fails with 415 from `MultipartSupport.findReader`, even when a `ParamConverter` for that type is registered. This happens both for a resource method parameter and for a field of a form class, since the converter is deliberately dropped as soon as a part media type is present.

Tested with a `ParamConverter` for a simple type and a part sent as `application/x-ticket`:

| case | result on main |
| --- | --- |
| `@RestForm Ticket` parameter, urlencoded and multipart | converter used |
| `@RestForm Ticket` field in a form class, urlencoded and multipart | converter used |
| `@RestForm @PartType("text/plain") Ticket` parameter | converter used |
| `@RestForm @PartType("application/x-ticket") Ticket` parameter | 415 |
| the same field in a form class | 415 |

Is the fallback still wanted? It would mean carrying the converter through to the multipart extractor and to the generated form class injection, and a request that returns 415 today would start succeeding when a converter exists. I am happy to open a PR for it, otherwise this issue looks resolved by #22205.
